backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers['on' + name]
+        const defaultHandler = defaultServerHandlers['on' + name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = Object.create(null)
+
+/**
+ * Emitted when a connection handshake completes.
+ *
+ * @param socket - The client socket that connected.
+ * @param request - The http GET request sent by the socket.
+ */
+defaultServerHandlers.onconnection = function onconnection (socket, request) {
+  socket.id = generateSocketID()
+  socket.isAlive = true
+  socket.server = this
+  socket.subscriptions = new Set()
+
+  const xForwardedFor = request.headers['x-forwarded-for']
+
+  const remoteAddress = (
+    typeof xForwardedFor === 'string' && xForwardedFor !== ''
+      ? xForwardedFor.split(/\s*,\s*/)[0]
+      : request.socket.remoteAddress
+  )
+  console.log(bold(`[pubsub] Socket ${socket.id} connected from ${remoteAddress}`))
+
+  // Create and attach socket event listeners.
+  ;['close', 'error', 'message', 'ping', 'pong'].forEach((eventName) => {
+    socket.on(eventName, (...args) => {
+      console.log(`[pubsub] Event '${eventName}' on socket ${socket.id}`, ...args)
+      const customHandler = socket.server.customClientHandlers['on' + eventName]
+      const defaultHandler = defaultClientHandlers['on' + eventName]
+
+      try {
+        if (defaultHandler) {
+          defaultHandler.call(socket, ...args)
```

Comment:
Well, its a trade-off between the `EventEmitter` and `EventTarget` interfaces.

When using an `EventEmitter`, the event target doesn't get passed through the `this` binding, making arrow functions a cleaner choice for event callbacks in pure Node code.

However when using an `EventTarget`, e.g. a WebSocket object in the browser, the socket does get passed through the `this` binding. Frontend developers are used to this behavior, so they will likely not use arrow functions as listeners there.

Since I had to implement both server-side and client-side components, I have chosen to favor the `EventTarget` interface to avoid using two different interfaces.
Other benefits may include a higher similarity between those components, and an hopefully an easier migration path to Deno.

Also, arrow functions are not needed to represent event handler maps when using shorthand method syntax.

Quick comparison:

1. Example using shorthand method syntax and socket smuggling:
```ts
const defaultClientHandlers = {
  onclose (code: string, reason: string) {
    ...
    this.subscriptions.clear()
  },
  onmessage (data: string) {
    ...
  }
}
```
2. Example using shorthand method syntax and explicit socket passing:
```ts
const defaultClientHandlers = {
  onclose (socket: Object, code: string, reason: string) {
    ...
    socket.subscriptions.clear()
  },
  onmessage (socket: Object, data: string) {
    ...
  }
}
```
3. Example using arrow functions. Not really nice here:
```ts
const defaultClientHandlers = {
  onclose: (socket: Object, code: string, reason: string) => {
    ...
    socket.subscriptions.clear()
  },
  onmessage: (socket: Object, data: string) => {
    ...
  }
}
```

Reply:
> Well, its a trade-off between the `EventEmitter` and `EventTarget` interfaces.
> 
> When using an `EventEmitter`, the event target doesn't get passed through the `this` binding, making arrow functions a cleaner choice for event callbacks in pure Node code.
> 
> However when using an `EventTarget`, e.g. a WebSocket object in the browser, the socket does get passed through the `this` binding. Frontend developers are used to this behavior, so they will likely not use arrow functions as listeners there.
> 
> Since I had to implement both server-side and client-side components, I have chosen to favor the `EventTarget` interface to avoid using two different interfaces.
> Other benefits may include a higher similarity between those components, and an hopefully an easier migration path to Deno.

Honestly, I don't think that emulating the event interface of browser APIs - which is notoriously annoying to work with - is a good goal to strive for, here. The broader JS ecosystem has essentially standardized on the `EventEmitter` approach and variants thereof, precisely because it has less weird design caveats and is much closer to how the rest of JS works semantically.

Plus, if we're going to try and make it consistent with the `EventTarget` design, then the current implementation would still not meet that bar - it's typical for browser-style event handlers to be called with a singular `Event` object that contains all the event data, rather than multiple arguments of varying types, as we are doing here. This is another reason that IMO the `EventEmitter` design is more useful.

(Speaking of, why *aren't* we just using Node's core EventEmitter? It should bundle fine.)

> Also, arrow functions are not needed to represent event handler maps when using shorthand method syntax.

Correct, but this doesn't hold when you start adding levels of abstraction. For example, consider:

```js
function createThing(options) {
	// implementation using the `this`-based API design for event handlers
	// ...
}

function createCustomThing(handler) {
	return createThing({
		handlers: {
			onStuff: handler
		}
	});
}

createCustomThing((eventArg) => {
	doWhatever(eventArg, 2)
});
```

In the `createCustomThing` invocation, it's not possible to use shorthand method syntax, because we're not in an object context. The same would be true if we first assign the handler to a variable, or process it in any way that isn't a literal specification within an object.

Reply:
> (Speaking of, why aren't we just using Node's core EventEmitter? It should bundle fine.)

This indeed could make the API more convenient, but I would rather avoid adding any new dependency to the frontend, given that one of the main goals of this PR is to drop dependencies.

Reply:
> given that one of the main goals of this PR is to drop dependencies.

I don't agree that this is a useful goal, stated as such. From what I recall, the problem with Primus was specifically that it is unreliable / unpleasant to work with, and difficult to support outside of JS, and those are definitely valid reasons to remove or avoid a dependency. It's the same reason I usually recommend people to rip out Socket.io before doing anything else.

But avoiding dependencies for the sake of being able to say that you're not using dependencies, while then reinventing them yourself, is not a net positive thing - at that point you're just moving from well-tested complexity to untested homegrown complexity. The complexity is *still there*, but now you're no longer benefiting from the shared work and experience of the broader community.

Dependencies are not a bad thing at all, you only need to make sure that you don't add dependencies which have higher costs than benefits (which is almost always the case for monolithic/kitchensink dependencies in particular, and rarely for modular dependencies). But Node's core EventEmitter is IMO more than reliable enough to depend on it, and it has a well-defined scope.

Reply:
On the topic of dependencies the goal of this PR: I agree with you @joepie91 that it doesn't make sense to re-invent the wheel for dependencies that work perfectly fine, which is why the issue title says to "adopt/fork rest". That makes it so that we can benefit from the collective labor without surprise updates that introduce vulnerabilities. With respect to Primus, the problems were:

- It required a very weird/ugly API to generate a client-side file that would randomly change periodically, making it difficult to be sure we weren't introducing vulnerabilities
- The size of this client-side file was monstrous, as it included dependencies for several other projects. By implementing our own solution we wouldn't have to duplicate existing dependencies, allowing us to keep the bundle size much smaller, speeding up loading speeds on slow connections
- Some of the Primus plugins were were using had outdated dependencies of their own with vulnerabilities that `npm` would complain about

And probably some others.

With respect to the EventEmitter API, I have no opinion, and will leave that up to your judgements.

Reply:
> which is why the issue title says to "adopt/fork rest". That makes it so that we can benefit from the collective labor without surprise updates that introduce vulnerabilities.

For what it's worth, forking isn't necessary here; a lockfile will already pin a dependency to a specific hash, and prevent automatic upgrades.

Reply:
Yeah, hopefully it works as advertised.

Reply:
By the way, if we wanted to keep Primus for its convenient API and other benefits, some of the associated problems could likely be mitigated by taking a few measures:
- The client-side Primus library file could be git-ignored so as not to disrupt git diffs.
- The Primus-related Grunt task could minimize that file to reduce the impact on the bundle size.
- It could also try to generate it only when it's outdated, not on every build.
- The `debug` dependency vulnerability could be solved by patching it ourselves or forcing Primus to use an up-to-date version.

It must be noted that all these measures would still require a non-trivial amount of work.

In my opinion, the best argument in favor of dropping Primus is the difficulty to make it work in other runtimes like Deno. Sure there is still a compatibility issue even when only using the `ws` library, but much smaller.

Reply:
@joepie91
> In the createCustomThing invocation, it's not possible to use shorthand method syntax, because we're not in an object context.

I'm not sure to understand correctly, because shorthand method syntax seems to work just fine e.g. in `createGIPubSubClient()` implemented in `~/frontend/controller/backend.js`:
```js
export function createGIPubSubClient (url: string, options: Object): Object {
  return createClient(url, {
    ...options,
    messageHandlers: {
      [NOTIFICATION_TYPE.ENTRY] (msg) {
        // We MUST use 'state/enqueueHandleEvent' here to ensure handleEvent()
        // is called AFTER any currently-running calls to syncContractWithServer().
        // Calling via SBP also makes it simple to implement 'test/backend.js'
        sbp('state/enqueueHandleEvent', GIMessage.deserialize(msg.data))
      }
    }
  })
}
```
Granted, it was not even necessary to use the `this` binding in the single message handler defined here, but I think it would have nonetheless be correctly bound to the client instance if used, namely because of the explicit binding via `handler.call(client)`.